src/foam/nanos/auth/AuthenticatedGroupDAO.java

```diff
@@ -0,0 +1,57 @@
+package foam.nanos.auth;
+
+import foam.core.FObject;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.dao.ProxyDAO;
+import foam.util.SafetyUtil;
+
+/**
+ * Custom authentication for groupDAO
+ *
+ * Features:
+ *   - If creating a group, you can only give it permissions implied by those that you already have.
+ *   - If updating a group, you can't modify the permissions array to have a permission that isn't implied by one you have.
+ */
+public class AuthenticatedGroupDAO extends ProxyDAO {
+  public AuthenticatedGroupDAO(X x, DAO delegate) {
+    super(x, delegate);
+  }
+
+  /**
+   * Make sure that the user has all permissions in the given group.
+   */
+  public void checkUserHasAllPermissionsInGroup(X x, Group toCheck) {
+    AuthService auth = (AuthService) x.get("auth");
+    String groupId = toCheck.getId();
+
+    while ( ! SafetyUtil.isEmpty(groupId) ) {
+      Group group = (Group) getDelegate().find_(x, groupId);
```

Comment:
The first time this loop runs the Group return will be equal to the Group that you passed in correct? It's minor but it would be more optimized if you eliminated the first call to the DAO.

Reply:
Yeah, on top of that this actually breaks creating a group too because the `find` will always return null for a new group. Fixing now.